rclcpp/src/rclcpp/node_interfaces/node_parameters.cpp

```diff
@@ -185,6 +188,102 @@ __lockless_has_parameter(
   return parameters.find(name) != parameters.end();
 }
 
+// see https://en.cppreference.com/w/cpp/types/numeric_limits/epsilon
+#define ULP 100
+RCLCPP_LOCAL
+bool
+__are_doubles_equal(double x, double y, int ulp = ULP)
+{
+  return std::abs(x - y) <= std::numeric_limits<double>::epsilon() * std::abs(x + y) * ulp;
+}
+
+#define FORMAT_REASON(reason, name, range_type) \
```

Comment:
I would suggest you make this a function instead of a macro.

Reply:
Done in https://github.com/ros2/rclcpp/pull/735/commits/b43d35c9647f8311ba0152b105f233de75586588